rllib/utils/exploration/gaussian_noise.py

```diff
@@ -0,0 +1,165 @@
+        gaussian_sample = self.scale_schedule(ts) * tf.random_normal(
+            tf.shape(deterministic_actions), stddev=self.stddev)
+
+        # Stochastic actions could either be: random OR action + noise.
+        random_actions, _ = \
+            self.random_exploration.get_tf_exploration_action_op(
+                action_dist, explore)
+        stochastic_actions = tf.cond(
+            pred=ts <= self.random_timesteps,
+            true_fn=lambda: random_actions,
+            false_fn=lambda: tf.clip_by_value(
+                deterministic_actions + gaussian_sample,
+                self.action_space.low * tf.ones_like(deterministic_actions),
+                self.action_space.high * tf.ones_like(deterministic_actions))
+        )
+
+        # Chose by `explore` (main exploration switch).
+        batch_size = tf.shape(deterministic_actions)[0]
+        action = tf.cond(
+            pred=tf.constant(explore, dtype=tf.bool)
+            if isinstance(explore, bool) else explore,
+            true_fn=lambda: stochastic_actions,
+            false_fn=lambda: deterministic_actions)
+        # Logp=always zero.
+        logp = tf.zeros(shape=(batch_size, ), dtype=tf.float32)
+
+        # Increment `last_timestep` by 1 (or set to `timestep`).
+        assign_op = \
+            tf.assign_add(self.last_timestep, 1) if timestep is None else \
+            tf.assign(self.last_timestep, timestep)
+        with tf.control_dependencies([assign_op]):
+            return action, logp
+
+    def _get_torch_exploration_action(self, action_dist, explore, timestep):
+        # Set last timestep or (if not given) increase by one.
+        self.last_timestep = timestep if timestep is not None else \
+            self.last_timestep + 1
+
+        # Apply exploration.
+        if explore:
+            # Random exploration phase.
+            if self.last_timestep <= self.random_timesteps:
+                action, _ = \
+                    self.random_exploration.get_torch_exploration_action(
+                        action_dist, True)
+            # Take a Gaussian sample with our stddev (mean=0.0) and scale it.
+            else:
+                det_actions = action_dist.deterministic_sample()
+                scale = self.scale_schedule(self.last_timestep)
+                gaussian_sample = scale * torch.normal(
+                    mean=0.0, stddev=self.stddev, size=det_actions.size())
+                action = torch.clamp(
+                    det_actions + gaussian_sample,
+                    self.action_space.low * torch.ones_like(det_actions),
+                    self.action_space.high * torch.ones_like(det_actions))
+        # No exploration -> Return deterministic actions.
+        else:
+            action = action_dist.deterministic_sample()
+
+        # Logp=always zero.
+        logp = torch.zeros(shape=(action.size()[0], ), dtype=torch.float32)
+
+        return action, logp
+
+    @override(Exploration)
+    def get_info(self):
```

Comment:
I think it makes more sense for this to be a separate `get_scale_value` method rather than overload the existing info method (which seems like it's only for debugging).

Reply:
Not sure I agree. Yes, right now, it's only used for recording this "info" (whatever it is for each Exploration class) in the train-results (Policy calls this generically and then tune json's it). This is useful, just like it is for current learning rate, memory consumption, etc.. (debugging/reporting snapshot of the state of the exploration object). I don't see any use for a specific `get_scale_value` method, as it would never be used by anyone (Policy should only use the top-level Exploration class methods, as it doesn't know, which specific sub-class it's holding).

Reply:
I was referring to the comment you had here:
```
            clean_actions, cur_noise_scale = self.sess.run(
                [self.output_actions,
                 self.exploration.get_info()],
```

It seemed odd to be passing self.exploration.get_info() to the session run, but if this is a temporary hack I guess it's fine. Otherwise, it seems more clear to assert the exploration is of type GaussianNoise, and then call a gaussian-noise specific method, rather than sneak the tensor through via get_info().

Reply:
Ah yes, absolutely. This is only a problem though right now as we don't have a parameter-noise exploration class yet and the parameter-noise logic piggybacks on that cur_noise_scale value (which it shouldn't, it should be separate exploration classes or a merged one). It's tagged as a `TODO(sven)` and will be fixed once we add ParameterNoise exploration classes.